# Анализ базы данных сервиса чтения книг по подписке



**Цель исследования -** выявить характеристики, которые определят самые популярные книги по рейтингам и писателям, а также обзорам от пользователей.

**Задачи:**
- проанализировать базу данных сервиса для чтения книг по подписке с информацией о книгах, издательствах, авторах и пользовательских обзорах
- посчитать, сколько книг вышло после 1 января 2000 года
- для каждой книги посчитать количество обзоров и среднюю оценку
- определить издательство, которое выпустило наибольшее число книг толще 50 страниц
- определить автора с самой высокой средней оценкой книг
- посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок
- сформулировать ценностное предложение для нового продукта.

**Описание данных:**

Таблица **books** - содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица **authors** - содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица **publishers** - содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

Таблица **ratings** - cодержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица **reviews** - cодержит данные о пользовательских обзорах:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

## Импорт библиотек и установка параметров

Для доступа к базе данных нам необходимо импортировать библиотеки и подключить некоторые параметры.

In [1]:
#импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
#устанавливаем параметры
db_config = {'user': 'praktikum_student', 
'pwd': 'Sdf4$2;d-d30pp', 
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, 
'db': 'data-analyst-final-project-db'} 
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
#сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Необходимые библиотеки импортированы, параметры для доступа к базе данных подключены.

## Исследование таблиц

Посмотрим содержание каждой таблицы, выведем первые строки.

### Таблица books

Содержит данные о книгах.

Имеет следующие поля:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Выведем первые 5 строк таблицы.

In [4]:
#выводим первые 5 строк таблицы books
query = '''
SELECT *
FROM books
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


Поля таблицы совпадают с описанием. В таблице 1000 строк.

### Таблица authors

Содержит данные об авторах.

Имеет следующие поля.
- `author_id` — идентификатор автора;
- `author` — имя автора.

Выведем первые 5 строк таблицы.

In [5]:
#выводим первые 5 строк таблицы authors
query = '''
SELECT *
FROM authors
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


Поля таблицы совпадают с описанием. В таблице 636 строк.

### Таблица publishers

Содержит данные об издательствах.

Поля таблицы:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

Выведем первые 5 строк таблицы.

In [6]:
#выводим первые 5 строк таблицы publishers
query = '''
SELECT *
FROM publishers
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


Поля таблицы совпадают с описанием. В таблице 340 строк.

### Таблица ratings

Содержит данные о пользовательских оценках книг.

Таблица содержит следующие поля:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Выведем первые 5 строк таблицы.

In [7]:
#выводим первые 5 строк таблицы ratings
query = '''
SELECT *
FROM ratings
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


Поля таблицы совпадают с описанием. В таблице 6456 строк.

### Таблица reviews

Содержит информацию о пользовательских обзорах.

В таблице присутствуют следующие поля:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

Выведем первые 5 строк таблицы.

In [8]:
#выводим первые 5 строк таблицы reviews
query = '''
SELECT *
FROM reviews
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


Поля таблицы совпадают с описанием. Таблица содержит 2793 строки.

## Исследовательский анализ данных

### Запрос №1

Посчитаем, сколько книг было выпущено после 1 января 2000 года.

In [9]:
#считаем книги, выпущенные после 01.01.2000
query = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,count
0,819


После 1 января 2000 года было выпущено 819 книг.

### Запрос №2

Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [10]:
#считаем количество обзоров и среднюю оценку для каждой книги
query = '''
SELECT b.title,
       COUNT(DISTINCT rev.review_id) AS review_cnt,
       AVG(rat.rating) AS average_rating
FROM books AS b
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
GROUP BY b.title
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,title,review_cnt,average_rating
0,'Salem's Lot,2,3.666667
1,1 000 Places to See Before You Die,1,2.500000
2,13 Little Blue Envelopes (Little Blue Envelope...,3,4.666667
3,1491: New Revelations of the Americas Before C...,2,4.500000
4,1776,4,4.000000
...,...,...,...
994,Wyrd Sisters (Discworld #6; Witches #2),3,3.666667
995,Xenocide (Ender's Saga #3),3,3.400000
996,Year of Wonders,4,3.200000
997,You Suck (A Love Story #2),2,4.500000


Мы вывели все книги и к каждой количество обзоров и средний рейтинг.

Больше всего обзоров у книги с названием Memoirs of a Geisha - 8 обзоров.

Книг с самым высоким рейтингом 5.0 44 штуки. Но среди таких книг есть одна, у которой также больше всего обзоров (4), -  A Dirty Job (Grim Reaper #1).

### Запрос №3

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц. Таким образом мы исключим из анализа брошюры.

In [11]:
#найдем издательство, которое выпустило наибольшее число книг толще 50 страниц
query = '''
SELECT DISTINCT p.publisher,
       COUNT(b.book_id) AS books_cnt
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY books_cnt DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,publisher,books_cnt
0,Penguin Books,42


Издательство Penguin Books выпустили самое большое количество книг толще 50 страниц - 42 книги.

### Запрос №4

Определим автора с самой высокой средней оценкой книг, у которых 50 и более оценок.

In [12]:
#найдем такого автора
query = '''
WITH author_rating AS (SELECT b.book_id,
                              b.author_id,
                              AVG(rat.rating) AS average_rating,
                              COUNT(rat.rating_id) AS rating_count
                       FROM books AS b
                       JOIN ratings AS rat ON b.book_id = rat.book_id
                       GROUP BY 1, 2
                       HAVING COUNT(rat.rating_id) >= 50)
SELECT a.author,
       AVG(ar.average_rating) AS avg_rating
FROM authors AS a
JOIN author_rating AS ar ON a.author_id = ar.author_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,author,avg_rating
0,J.K. Rowling/Mary GrandPré,4.283844


Автором с самым высоким средним рейтингов, среди книг с 50 и более оценками, оказалась J.K. Rowling и др. Средний рейтинг - 4,28.

### Запрос №5

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [13]:
query = '''
WITH big_rating AS (SELECT username,
                           COUNT(rating_id) AS rating_cnt
                    FROM ratings
                    GROUP BY 1
                    HAVING COUNT(rating_id) > 48),
     reviews_quantity AS (SELECT username,
                                 COUNT(review_id) AS review_cnt
                          FROM reviews
                          GROUP BY 1)
SELECT AVG(rq.review_cnt) AS avg_reviews
FROM big_rating AS br
LEFT JOIN reviews_quantity AS rq ON br.username = rq.username
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,avg_reviews
0,24.0


В среднем пользователи, которые поставили больше 48 оценок, оставляют 24 обзора на книги.

## Выводы

Нами был произведен анализ базы данных книжного сервиса. Нам удалось выявить следующее:
- В сервисе представлено 1000 книг, 636 авторов и 340 издательств.
- Пользователи поставили книгам 6456 оценок и написали 2793 обзора на книги.
- С 1 января 2000 года на сервисе было выпущено 819 книг, что составляет 81,9%.
- Наибольшее количество обзоров пользователи оставили для книги Memoirs of a Geisha - 8 обзоров.
- Самый высокий книжный рейтинг, полученный книгами, на сервисе - 5.0. Книг с таким рейтингом 44. Среди них есть книга с наибольшим количеством обзоров (4 обзора) - A Dirty Job (Grim Reaper #1).
- Самые толстые книги чаще всего выпускаются издательством Penguin Books. Издательство выпустило 42 книги с количеством страниц более 50. Это наибольшее количество среди всех издательств.
- В среднем самым рейтинговым автором, среди книг с 50 или более оценками, стала J.K. Rowling. Её средний рейтинг на сервисе составил 4,28.
- Пользователи, которые ставят более 48 оценок, в среднем пишут 24 обзора на книги.

**Рекомендации.**
Скорее всего мы имеем совсем небольшой книжный сервис, исходя из количества представленных на нем книг (всего 1000 штук) и наибольшему количеству обзоров у достаточно культовой книги "Мемуары гейши" (всего 8 штук). Также, скорее всего, пользователи сервиса достаточно лояльны, что можем видеть по среднему количеству обзоров пользователей, поставивших более 48 оценок.

Книжному сервису стоит смотреть в сторону расширения количества книг, таким образом на сервисе появятся новые, интересные читателям, жанры и авторы, что может помочь привлечь новую аудиторию пользователей.